### ISSUE TITLE
[BUG] Wrong energies with intel pair style

### ISSUE THREAD

**Issue by ab5424** [lammps/lammps#2892]
**Summary**

The pair style `buck/coul/long/intel` reports wrong energies, which can lead to inaccurate calculations.

**LAMMPS Version and Platform**

29Oct2020 & 30Jul2021 on CentOS

**Expected Behavior**

The total energy (`etotal`) should match between the `log.lammps` file and the `fix print` command.

**Actual Behavior**

When using `buck/coul/long/intel`, the `fix print` command writes out `elong` instead of `etotal` (when `etotal` is requested). The energies in the logfile are correct, and the `fix atom/swap` works as expected. When using an intel kspace solver in addition to the intel pair style, the `atom/swap` fix is inaccurate, as all swaps are accepted. I assume that in the latter case the wrong energies are considered for the swap.

**Steps to Reproduce**

Run a simulation with buck/coul/long and

1. no additional commands.
2. the intel pair style version (`buck/coul/long/intel`).
3. the intel pair style version (`buck/coul/long/intel`) and an intel kspace solver.

Write out energies with the `fix print` command and swap some atoms with `fix atom/swap`.

**Further Information, Files, and Links**

Example calculations can be found [here](https://gigamove.rwth-aachen.de/de/download/f7e86f30cd1b6626173d51c836f2a588) (for all three cases). The command for execution was `mpirun -np 8 lmp -in *.run.in -pk intel 0 omp 1 lrt yes` in all cases.

**Comment by akohlmey** [comment]
@ab5424 please clarify. Is using fix atom/swap a prerequisite to observe this issue or will the inconsistency also happen without?

**Comment by ab5424** [comment]
@akohlmey I didn't observe it when using fix npt or no fix at all.

**Comment by ab5424** [comment]
I also tried putting fix print before fix atom/swap but there were no changes.

**Comment by akohlmey** [comment]
> @akohlmey I didn't observe it when using fix npt or no fix at all.

That is what I suspected. Most of the INTEL package pair styles implement a caching of force field parameters in data structures that are required for efficient vectorization. Thus any fix that changes parameters may have to re-initialize those. This is usually done by the fix in question. Some fixes, e.g. fix adapt call such a function in the Pair class to trigger this (which could also make sing INTEL package pair styles significantly less efficient), because the same kind of procedure would be needed to recompute some intermediate per-atom type values. I suspect fix atom/swap does not need to do this for "normal" pair styles (and it also predates operations like what fix adapt does and the added APIs because of that) and thus force field settings or data structures in the /intel pair styles become inconsistent. There is a black-list somewhere to block the use of features in LAMMPS that are incompatible with the INTEL package optimizations, but that list definitely not complete since there are so many features in LAMMPS (and new features get added all the time) so that it is near impossible to watch out and check all of them.

I will have to have a closer look to either confirm my hypothesis or defer further investigation to one of the INTEL package developers. If my hunch is correct, then the solution would be to not use the /intel pair styles with fix atom/swap, unfortunately.

**Comment by akohlmey** [comment]
> @akohlmey Thanks for the detailed explanation. But would this also explain why the swapping (seemingly) works when using the intel pair style without the intel kspace solver? At least the acceptance rates for the swaps are very close to the calculation with the "normal" pair style.

No, this would be a general problem. It is difficult to say something specific without looking at the code and running some traces with valgrind to determine if there are some memory accesses using "wild pointers" causing the issue or something else.

> Would it be an option to use whitelisting instead/additionally? Or print some kind of warning?

This is too generic an issue. With whitelisting you would have to enable hundreds of fixes that have not been checked but have no problem. Can you imagine the flood of people complaining "Why does this not work with the INTEL package?"?. How to go about this was ultimately a decision of the INTEL package developers.
